local-interchaintest/src/tests/swap/token_swap.rs

```diff
@@ -0,0 +1,478 @@
+        splitter_address.clone(),
+        party_a_router_address.clone(),
+        party_b_router_address.clone(),
+    ];
+    if !party_a_ibc_forwarder_address.is_empty() {
+        addresses.push(party_a_ibc_forwarder_address.clone());
+    }
+    if !party_b_ibc_forwarder_address.is_empty() {
+        addresses.push(party_b_ibc_forwarder_address.clone());
+    }
+    for address in &addresses {
+        send(
+            test_ctx
+                .get_request_builder()
+                .get_request_builder(NEUTRON_CHAIN_NAME),
+            DEFAULT_KEY,
+            address,
+            &[Coin {
+                denom: neutron_denom.clone(),
+                amount: Uint128::new(5000000000),
+            }],
+            &Coin {
+                denom: neutron_denom.clone(),
+                amount: Uint128::new(5000),
+            },
+        )
+        .unwrap();
+    }
+
+    info!("Tick until forwarders create ICA...");
+    let party_a_deposit_address;
+    let party_b_deposit_address;
+    loop {
+        tick(
+            test_ctx
+                .get_request_builder()
+                .get_request_builder(NEUTRON_CHAIN_NAME),
+            DEFAULT_KEY,
+            &clock_address,
+        );
+        let forwarder_a_state = query_contract_state(
+            test_ctx
+                .get_request_builder()
+                .get_request_builder(NEUTRON_CHAIN_NAME),
+            &party_a_ibc_forwarder_address,
+        );
+        info!("Forwarder A state: {:?}", forwarder_a_state);
+        if forwarder_a_state == "ica_created" {
+            party_a_deposit_address = covenant.query_deposit_address("party_a".to_string());
+            party_b_deposit_address = covenant.query_deposit_address("party_b".to_string());
+            break;
+        }
+    }
+
+    info!("Party A deposit address: {}", party_a_deposit_address);
+    info!("Party B deposit address: {}", party_b_deposit_address);
+
+    info!("Fund the forwarders with sufficient funds...");
+    send(
+        test_ctx
+            .get_request_builder()
+            .get_request_builder(GAIA_CHAIN_NAME),
+        DEFAULT_KEY,
+        &party_a_deposit_address,
+        &[Coin {
+            denom: atom_denom.clone(),
+            amount: Uint128::new(uatom_contribution_amount),
+        }],
+        &Coin {
+            denom: atom_denom.clone(),
+            amount: Uint128::new(5000),
+        },
+    )
+    .unwrap();
+    send(
+        test_ctx
+            .get_request_builder()
+            .get_request_builder(NEUTRON_CHAIN_NAME),
+        DEFAULT_KEY,
+        &party_b_deposit_address,
+        &[Coin {
+            denom: neutron_denom.clone(),
+            amount: Uint128::new(untrn_contribution_amount),
+        }],
+        &Coin {
+            denom: neutron_denom.clone(),
+            amount: Uint128::new(5000),
+        },
+    )
+    .unwrap();
+
+    info!("Tick until forwarders forward the funds to the holder...");
+    loop {
```

Comment:
Should there be a timer to break out of the loop in case the funds are, for some reason, not forwarded to the holder.

Reply:
It won't ever happen, if it does it means the covenants are broken. We would be adding this break out and error logic in many places instead of letting it just time out, not sure about it. Curious what @bekauz things.

Reply:
Yeah it currently just times out the CI and fails. It would be nice to be able to detect dead loops, but I'm not sure if it's high priority given that our e2e tests are happy-case only?